### PR TITLE
chore(ios): move application shortcut under Ti.UI.Shortcut to have parity

### PR DESCRIPTION
**Pull request by vijaysingh-axway**
https://jira.appcelerator.org/browse/TIMOB-26818

NOTE - Merge with PR #11759

### CODE REVIEW

**Review by sgtcoolguy — CHANGES_REQUESTED** [review]
I think we definitely need some basic unit tests around this.

Probably need to take this: https://github.com/appcelerator/titanium-mobile-mocha-suite/blob/master/Resources/ti.ui.shortcutitem.test.js

Add it to `tests/Resources` here and drop the `.android` filters on the suite/tests.

File: iphone/Classes/TiUIShortcutProxy.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+* Please see the LICENSE included with this distribution for details.
+*/
+
+#if defined(USE_TI_UISHORTCUT) || defined(USE_TI_UISHORTCUTITEM)
+
+#import <TitaniumKit/TitaniumKit.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol TiUIShortcutProxyExports <JSExport>
```
[the remark this answers is not in the document]
Comment (vijaysingh-axway) [reply iphone/Classes/TiUIShortcutProxy.h:16]: :)

File: iphone/Classes/TiUIShortcutProxy.m [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+- (void)add:(id)arg
+{
+  ENSURE_SINGLE_ARG(arg, TiUIShortcutItemProxy);
+
+  NSMutableArray *shortcuts = (NSMutableArray *)[UIApplication sharedApplication].shortcutItems;
+
+  // remove the previous shortcutitem of same id if exists
+  for (UIApplicationShortcutItem *item in shortcuts) {
+    if ([item.type isEqualToString:[((TiUIShortcutItemProxy *)arg) shortcutItem].type]) {
```
[the remark this answers is not in the document]
Comment (vijaysingh-axway) [reply iphone/Classes/TiUIShortcutProxy.m]: In native iOS, there is 'type' property which we used to use uniquely identify the shortcut. To make it compatible with android this is exposed as 'id'.

File: iphone/Classes/TiUIShortcutProxy.m [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+- (void)remove:(id)arg
+{
+  ENSURE_SINGLE_ARG(arg, TiUIShortcutItemProxy);
+
+  NSString *key = [((TiUIShortcutItemProxy *)arg) shortcutItem].type;
+
+  NSMutableArray *shortcuts = (NSMutableArray *)[UIApplication sharedApplication].shortcutItems;
+  for (UIApplicationShortcutItem *item in shortcuts) {
+    if ([item.type isEqualToString:[((TiUIShortcutItemProxy *)arg) shortcutItem].type]) {
```
[the remark this answers is not in the document]
Comment (vijaysingh-axway) [reply iphone/Classes/TiUIShortcutProxy.m]: In native iOS, there is 'type' property which we used to use uniquely identify the shortcut. To make it compatible with android this is exposed as 'id'.

File: iphone/Classes/TiUIShortcutProxy.m [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  return shortcutsToReturn;
+}
+
+- (TiUIShortcutItemProxy *)getById:(id)identifier
+{
+  ENSURE_SINGLE_ARG(identifier, NSString);
+
+  NSArray *shortcuts = [UIApplication sharedApplication].shortcutItems;
+  for (UIApplicationShortcutItem *item in shortcuts) {
+    if ([item.type isEqualToString:[TiUtils stringValue:identifier]]) {
```
[the remark this answers is not in the document]
Comment (vijaysingh-axway) [reply iphone/Classes/TiUIShortcutProxy.m:102]: In native iOS, there is 'type' property which we used to use uniquely identify the shortcut. To make it compatible with android this is exposed as 'id'.

File: iphone/Classes/TiUIShortcutItemProxy.m [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+* Licensed under the terms of the Apache Public License
+* Please see the LICENSE included with this distribution for details.
+*/
+#if defined(USE_TI_UISHORTCUT) || defined(USE_TI_UISHORTCUTITEM)
+
+#import "TiUIShortcutItemProxy.h"
+#ifdef USE_TI_CONTACTS
+#import "TiContactsPerson.h"
+#endif
+#import <ContactsUI/ContactsUI.h>
```
[the remark this answers is not in the document]
Comment (vijaysingh-axway) [reply iphone/Classes/TiUIShortcutItemProxy.m:14]: Yes, it is needed. Moved under USE_TI_CONTACTS guard.

File: iphone/Classes/TiUIiOSProxy.m [hunk 6]
```diff
@@ -750,6 +750,7 @@ - (id)createDynamicItemBehavior:(id)args
 #ifdef USE_TI_UIIOSAPPLICATIONSHORTCUTS
 - (id)createApplicationShortcuts:(id)args
 {
+  DEPRECATED_REPLACED(@"UI.iOS.ApplicationShortcuts", @"9.1.0", @"UI.Shortcut");
   return [[[TiUIiOSApplicationShortcutsProxy alloc] _initWithPageContext:[self executionContext] args:args] autorelease];
```
[the remark this answers is not in the document]
Comment (vijaysingh-axway) [reply iphone/Classes/TiUIiOSProxy.m:754]: This proxy, TiUIiOSProxy, is not ObjcProxy. So it should be fine.

**Comment by vijaysingh-axway** [comment]
> I think we definitely need some basic unit tests around this.
> 
> Probably need to take this: https://github.com/appcelerator/titanium-mobile-mocha-suite/blob/master/Resources/ti.ui.shortcutitem.test.js
> 
> Add it to `tests/Resources` here and drop the `.android` filters on the suite/tests.

I have seen Gary has written unit test case in his PR. I tried to run at my end , it is working fine. Once that PR get merge, I'll update that to run for iOS as well.

File: iphone/Classes/TiUIShortcutProxy.m [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+- (void)removeAll
+{
+  [UIApplication sharedApplication].shortcutItems = nil;
+}
+
+- (void)add:(TiUIShortcutItemProxy *)shortcut
+{
+  NSMutableArray<UIApplicationShortcutItem *> *shortcuts = (NSMutableArray<UIApplicationShortcutItem *> *)[UIApplication sharedApplication].shortcutItems;
+
+  // remove the previous shortcutitem of same id if exists
```
Comment (sgtcoolguy) [inline iphone/Classes/TiUIShortcutProxy.m:133]: do we need to worry about position? In other words, should we replace the existing shortcut with that id so we retain the same index/position in the array? or does position not matter here?

File: iphone/Classes/TiUIShortcutProxy.h [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+READONLY_PROPERTY(NSArray<TiUIShortcutItemProxy *> *, items, Items);
+READONLY_PROPERTY(NSArray<TiUIShortcutItemProxy *> *, staticItems, StaticItems);
+
+- (TiUIShortcutItemProxy *)getById:(NSString *)identifier;
+
+- (void)remove:(TiUIShortcutItemProxy *)shortcut;
```
Comment (sgtcoolguy) [inline iphone/Classes/TiUIShortcutProxy.h:22]: cc @garymathews So this is me really super nit-picking. Personally I always do a double-take on any method that returns void. It offers no direct means of determining what happened, or if something happened. You need to usually poke around the object more to tell if what you wanted to happen actually did.

So in this case, we're really talking about a collection and simply adding/removing elements. I think it'd be useful to know if that actually happened. So I think perhaps in this case a boolean return type may make sense.

For example, what if I pass in a shortcut to remove that was never added in the first place? That should return false and do nothing. What if the collection was already empty, removeAll should return false. Add would likely always return true or stay void - any inability to add a shortcut should more likely be an Error thrown.

You guys don't need to change this, just something to consider. Languages/frameworks are very spotty for something like these operations where a lot return void (Obj-C NSMutableArray), some return booleans (Java List), JS returns the new length of the collection for Array. So it's not a clear consensus. (And in Java's case, `add` always returns `true`)

File: iphone/Classes/TiUIShortcutProxy.m (hunk 7 above)
Comment (vijaysingh-axway) [reply iphone/Classes/TiUIShortcutProxy.m]: Good catch! Position matters. Updated to maintain position.

**Review by sgtcoolguy — COMMENTED** [review]
@vijaysingh-axway  Please copy the tests from @garymathews 's PR for Android (and make them also run on iOS).

**Comment by vijaysingh-axway** [comment]
> @vijaysingh-axway Please copy the tests from @garymathews 's PR for Android (and make them also run on iOS).

@sgtcoolguy Added the tests. 
'remove' test case has been updated a bit. Reason is that 'remove' test case get finished before 'add'. In that case shortcut count get mismatched. While resolving merge conflict 'remove' test case from this PR need to be taken.

**Review by sgtcoolguy — CHANGES_REQUESTED** [review]
Just needs updated apidocs

File: iphone/Classes/TiUIShortcutItemProxy.h [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+#if defined(USE_TI_UISHORTCUT) || defined(USE_TI_UISHORTCUTITEM)
+#import <TitaniumKit/ObjcProxy.h>
+
+NS_ASSUME_NONNULL_BEGIN
+@protocol TiUIShortcutItemProxyExports <JSExport>
+
+READONLY_PROPERTY(NSString *, title, Title);
+READONLY_PROPERTY(NSString *, description, Description);
+READONLY_PROPERTY(NSDictionary *, data, Data);
+READONLY_PROPERTY(id, icon, Icon);
```
[the remark this answers is not in the document]
Comment (vijaysingh-axway) [reply iphone/Classes/TiUIShortcutItemProxy.h:16]: Updated doc!

File: iphone/Classes/TiUIShortcutItemProxy.m [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+  }
+
+#ifdef USE_TI_CONTACTS
+  if ([value isKindOfClass:[TiContactsPerson class]]) {
+    ENSURE_TYPE(value, TiContactsPerson);
+    return [UIApplicationShortcutIcon iconWithContact:[(TiContactsPerson *)value nativePerson]];
+  }
+#endif
+
+  if ([value isKindOfClass:[UIApplicationShortcutIcon class]]) {
```
[the remark this answers is not in the document]
Comment (vijaysingh-axway) [reply iphone/Classes/TiUIShortcutItemProxy.m]: There is no way developer can pass this. Removed it.

File: iphone/Classes/TiUIShortcutItemProxy.m [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+      NSLog(@"[ERROR] Ti.UI.ShortcutItem: The \"id\" property is required.");
+      return;
+    }
+
+    if ([args valueForKey:@"title"] == nil) {
+      NSLog(@"[ERROR] Ti.UI.ShortcutItem: The \"title\" property is required.");
+      return;
+    }
+
+    _shortcutItem = [[UIApplicationShortcutItem alloc] initWithType:[args valueForKey:@"id"]
```
[the remark this answers is not in the document]
Comment (vijaysingh-axway) [reply iphone/Classes/TiUIShortcutItemProxy.m]: Updated. Thanks!

File: iphone/Classes/TiUIShortcutItemProxy.h [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+*/
+#if defined(USE_TI_UISHORTCUT) || defined(USE_TI_UISHORTCUTITEM)
+#import <TitaniumKit/ObjcProxy.h>
+
+NS_ASSUME_NONNULL_BEGIN
+@protocol TiUIShortcutItemProxyExports <JSExport>
+
+READONLY_PROPERTY(NSString *, title, Title);
+READONLY_PROPERTY(NSString *, description, Description);
+READONLY_PROPERTY(NSDictionary *, data, Data);
```
[the remark this answers is not in the document]
Comment (vijaysingh-axway) [reply iphone/Classes/TiUIShortcutItemProxy.h:15]: It should Dictionary.  Updated doc.

**Comment by vijaysingh-axway** [comment]
@garymathews In android, As per  [doc] (https://github.com/appcelerator/titanium_mobile/blob/a378d21e0e5e2343ea3617941226066d8b48321e/apidoc/Titanium/UI/Shortcut.yml#L53), callback of Ti.UI.Shortcut `click` event should give 'item' property which should be Ti.UI.ShortcutItem, which is expected behavior as well.
But in example code [here](https://github.com/appcelerator/titanium_mobile/blob/a378d21e0e5e2343ea3617941226066d8b48321e/apidoc/Titanium/UI/Shortcut.yml#L69), you are using 'e.id' instead of 'e.item.id' to log shortcut id. Is this documentation issue or android has implemented in this way?
cc @sgtcoolguy @jquick-axway .

**Comment by garymathews** [comment]
> @garymathews In android, As per [doc] (
> 
> https://github.com/appcelerator/titanium_mobile/blob/a378d21e0e5e2343ea3617941226066d8b48321e/apidoc/Titanium/UI/Shortcut.yml#L53
> 
> ), callback of Ti.UI.Shortcut `click` event should give 'item' property which should be Ti.UI.ShortcutItem, which is expected behavior as well.
> But in example code [here](https://github.com/appcelerator/titanium_mobile/blob/a378d21e0e5e2343ea3617941226066d8b48321e/apidoc/Titanium/UI/Shortcut.yml#L69), you are using 'e.id' instead of 'e.item.id' to log shortcut id. Is this documentation issue or android has implemented in this way?
> cc @sgtcoolguy @jquick-axway .

Good catch, I've amended this with https://github.com/appcelerator/titanium_mobile/pull/11823

File: apidoc/Titanium/UI/Shortcut.yml [hunk 13]
```diff
@@ -59,14 +63,14 @@ events:
 examples:
   - title: Simple Shortcut Example
     example: |
-        Create a shortcut item and detect its click with the `shortcutitemclick` event.
+        Create a shortcut item and detect its click with the `click` event.
 
-        ``` javascript
+        ```
```
Comment (sgtcoolguy) [inline apidoc/Titanium/UI/Shortcut.yml:68]: we want to declare the code blocks' language as we did before (it lets the doc sites syntax color the code properly)

Comment (vijaysingh-axway) [reply apidoc/Titanium/UI/Shortcut.yml]: fixed.

**Comment by sgtcoolguy** [comment]
manually rebased, squashed down commits, combined with latest unit tests for API and merged to master

**Pull request closed without merging**